### ISSUE TITLE
Detail of c++/simupop implementations

### ISSUE THREAD

**Issue by molpopgen** [petrelharp/ftprime_ms#33]
Question for the group:

Do we want an "Implementation details" section describing more of the gory details?  Happy to do it, and describing the quirks of how the threaded version works may be of some interest?

**Comment by petrelharp** [comment]
I am unsure.  I think we are running the risk of having too many gory
details (eg, the simplify algorithm as currently written); but then I think
some of these are useful.  If you feel compelled to write it, we could
always put it in the "Methods" section.

I think that details relevant to any implementation are useful (like, only
one-way communication is needed), but ones specific to this implementation
(maybe, the GIL?) are less so?

**Comment by jeromekelleher** [comment]
I'm with Peter here. High-level stuff that would be of use to someone else implementing a simulator based on this framework would be interesting, but intimate details about a specific implementation less so.

**Comment by molpopgen** [comment]
OK, I'm in agreement there.  

This segues another issue, which is the existing documentation.  One of the challenges of implementing this stuff is that it requires knowing the precise `dtype` of each element in the relevant tables, etc.  Those fields are currently not documented.  I'm fine adding details to the README for my implementation, but where do we expect msprime to be by the time we submit?

**Comment by jeromekelleher** [comment]
> This segues another issue, which is the existing documentation. One of the challenges of implementing this stuff is that it requires knowing the precise dtype of each element in the relevant tables, etc. Those fields are currently not documented. I'm fine adding details to the README for my implementation, but where do we expect msprime to be by the time we submit?

The tables API will be fully documented and released in 0.5.0. This should be done before the Christmas break. I don't think this should be a blocker for submitting/putting on Biorxiv though. What do others think? Is msprime 0.5.0 a pre-requisite for submitting the paper?

**Comment by ashander** [comment]
> The tables API will be fully documented and released in 0.5.0. This should be done before the Christmas break. I don't think this should be a blocker for submitting/putting on Biorxiv though. What do others think? Is msprime 0.5.0 a pre-requisite for submitting the paper?

Sounds like we'll have 0.5.0 before publication in any case and can add that to the published version. (Is it still the case that one _cannot_ update / version a biorxiv paper?)  I'm fine with submitting before 0.5.0 is actually released so long as @jeromekelleher is too.

**Comment by molpopgen** [comment]
This sounds good.   I bring it up because I've handled papers as an editor and as a reviewer where the software was not readily installable/usable by the reviewers.

**Comment by jeromekelleher** [comment]
> This sounds good. I bring it up because I've handled papers as an editor and as a reviewer where the software was not readily installable/usable by the reviewers.

That's a fair point, I hadn't thought about the actual review process! OK, let's push 0.5.0 up the priority list a bit.

**Comment by petrelharp** [comment]
Things currently run correctly, so I vote for not letting documentation
hold up bioRxiv/submission, but making an issue for documentation and
putting a note up top of the README that the documentation is in progress,
so it's clear that it's happening and folks have someplace to ask for
help.  And prioritizing writing the documentation after that!

**Comment by molpopgen** [comment]
One intermediate step may be a unit test file that creates instance of all the table types and then asserts what their `dtype`s are.  This would serve as a rudimentary documentation and also enforce some future-proofing onto the API in that it cannot be changed without @jeromekelleher annoying himself each time.

**Comment by jeromekelleher** [comment]
There is already a bunch of unit tests on this stuff @molpopgen, which are reasonably declarative:

https://github.com/jeromekelleher/msprime/blob/master/tests/test_tables.py#L407

Not that this is a replacement for documentation. We're certainly not going to be changing the dtypes once 0.5.0 is released though, I want this to be a long-term stable API.

**Comment by petrelharp** [comment]
Have a look at the Methods and see if there's any more useful gory details to put in there?  Currently there's nothing about the 'parallel' version.

**Comment by molpopgen** [comment]
> Have a look at the Methods and see if there's any more useful gory details to put in there? Currently there's nothing about the 'parallel' version.

I'll take a look.   There is a brief description in the figure caption, though, which would simply be replicated in the Methods.

**Comment by petrelharp** [comment]
If the details in the figure caption are sufficient, then no need to replicate them! I recalled you working through some more things to make it work; wasn't sure if you wanted to describe those.

**Comment by molpopgen** [comment]
I submitted a PR that basically just gives the names of what I used.  There's a wacky implementation trick, which is a bit too out there for the paper, involving keeping memory pre-allocated for each trip into the queue.  I'm not super-convinced that keeping the memory around is important beyond the notion that it is a bit more pleasing to keep big chunks of memory around if you know you're gonna reuse them.
